tests/test_fipa_end2end.py

```diff
@@ -0,0 +1,307 @@
+
+import pytest
+
+from aea.aea_builder import AEABuilder
+from aea.common import Address
+from aea.configurations.base import SkillConfig
+from aea.helpers.base import cd
+from aea.helpers.search.models import Constraint, ConstraintType, Description, Query
+from aea.protocols.base import Message
+from aea.protocols.dialogue.base import Dialogue as BaseDialogue
+from aea.skills.base import Handler, Skill, SkillContext
+from aea.skills.behaviours import TickerBehaviour
+from aea.test_tools.test_cases import AEATestCaseEmpty
+
+from packages.fetchai.protocols.fipa.dialogues import FipaDialogue, FipaDialogues
+from packages.fetchai.protocols.fipa.message import FipaMessage
+
+from tests.common.utils import run_in_thread, wait_for_condition
+
+
+class SellerDialogues(FipaDialogues):
+    """The dialogues class keeps track of all dialogues."""
+
+    def __init__(self, self_address: Address) -> None:
+        """
+        Initialize dialogues.
+
+        :return: None
+        """
+
+        def role_from_first_message(  # pylint: disable=unused-argument
+            message: Message, receiver_address: Address
+        ) -> BaseDialogue.Role:
+            """Infer the role of the agent from an incoming/outgoing first message
+
+            :param message: an incoming/outgoing first message
+            :param receiver_address: the address of the receiving agent
+            :return: The role of the agent
+            """
+            return FipaDialogue.Role.SELLER
+
+        FipaDialogues.__init__(
+            self,
+            self_address=self_address,
+            role_from_first_message=role_from_first_message,
+            dialogue_class=FipaDialogue,
+        )
+
+
+class BuyerDialogues(FipaDialogues):
+    """The dialogues class keeps track of all dialogues."""
+
+    def __init__(self, self_address: Address) -> None:
+        """
+        Initialize dialogues.
+
+        :return: None
+        """
+
+        def role_from_first_message(  # pylint: disable=unused-argument
+            message: Message, receiver_address: Address
+        ) -> BaseDialogue.Role:
+            """Infer the role of the agent from an incoming/outgoing first message
+
+            :param message: an incoming/outgoing first message
+            :param receiver_address: the address of the receiving agent
+            :return: The role of the agent
+            """
+            return FipaDialogue.Role.BUYER
+
+        FipaDialogues.__init__(
+            self,
+            self_address=self_address,
+            role_from_first_message=role_from_first_message,
+            dialogue_class=FipaDialogue,
+        )
+
+
+class BuyerBehaviour(TickerBehaviour):
+    """Test buyer behaviour."""
+
+    addr: str
+
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.is_started = False
+        self.was_called = False
+
+    def act(self) -> None:
+        """Make an action."""
+        self.dialogues = self.context.shared_state["dialogues"]
```

Comment:
Dialogues should not be stored in shared_state. Just create in init! The shared_state is for cross-skill usage.

Reply:
updated